tests/gauge_force_test.cpp

```diff
@@ -214,6 +236,127 @@ void gauge_force_test(bool compute_force = true)
   delete Mom_ref_milc;
 }
 
+void gauge_loop_test()
+{
+  int max_length = 6;
+
+  setVerbosityQuda(QUDA_VERBOSE, "", stdout);
+
+  QudaGaugeParam gauge_param = newQudaGaugeParam();
+  setGaugeParam(gauge_param);
+
+  gauge_param.gauge_order = gauge_order;
+  gauge_param.t_boundary = QUDA_PERIODIC_T;
+
+  setDims(gauge_param.X);
+
+  double *trace_loop_coeff_p = &trace_loop_coeff_d[0];
+  int num_paths = sizeof(trace_path) / sizeof(trace_path[0]);
+  int *trace_loop_length_p = &trace_loop_length[0];
+
+  // b/c we can't cast int*[6] -> int**
+  int** trace_path_p = new int*[num_paths];
+  for (int i = 0; i < num_paths; i++) {
+    trace_path_p[i] = new int[max_length];
+    for (int j = 0; j < trace_loop_length[i]; j++)
+      trace_path_p[i][j] = trace_path[i][j];
+  }
+
+  quda::GaugeFieldParam param(gauge_param);
+  param.create = QUDA_NULL_FIELD_CREATE;
+  param.order = QUDA_QDP_GAUGE_ORDER;
+  auto U_qdp = new quda::cpuGaugeField(param);
+
+  // fills the gauge field with random numbers
+  createSiteLinkCPU((void **)U_qdp->Gauge_p(), gauge_param.cpu_prec, 0);
+
+  param.order = QUDA_MILC_GAUGE_ORDER;
+  auto U_milc = new quda::cpuGaugeField(param);
+  if (gauge_order == QUDA_MILC_GAUGE_ORDER) U_milc->copy(*U_qdp);
+
+  void *sitelink = nullptr;
+
+  if (gauge_order == QUDA_MILC_GAUGE_ORDER) {
+    sitelink = U_milc->Gauge_p();
+  } else if (gauge_order == QUDA_QDP_GAUGE_ORDER) {
+    sitelink = U_qdp->Gauge_p();
+  } else {
+    errorQuda("Unsupported gauge order %d", gauge_order);
+  }
+
+  // storage for traces
+  using double_complex = double _Complex;
+  double_complex* traces = new double_complex[num_paths];
```

Comment:
replace with a `std::vector<double_complex>` to avoid needing to free at the end

Reply:
Is it legal to pass that as a pointer in as `&traces[0]`? I think I ran into some nightmares with passing different flavors of complex numbers (C vs C++ vs ...) into `computeGaugeLoopTraceQuda`, but that could be revisited. I could also just follow the convention I have elsewhere of passing in an array/vector of doubles of length 2 * loops.

Reply:
C and C++ complex numbers are defined by standard to be compatible, and even directly "castable" to each other.  This is one specific place where `reinterpret_cast` as opposed to `memcpy` is preferable.

To pass it as a pointer, one should use `traces.data()` as opposed to `&traces[0]`, with the latter indeed, likely being UB.

Reply:
https://github.com/lattice/quda/pull/1308/commits/1e3db7ab2f46d71cf3b88aa7e904b8b4f9c8c834